File: 59.spiral-matrix-ii.js
/*
 * @lc app=leetcode id=59 lang=javascript
 *
 * [59] Spiral Matrix II
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function (n) {
  const res = Array(n)
    .fill()
    .map(() => Array(n));
  let val = 1;
  let [left, right, top, bottom] = [0, n - 1, 0, n - 1];
  while (left <= right && top <= bottom) {
    for (let i = left; i < right + 1; ++i) {
      res[top][i] = val++;
    }
    if (++top > bottom) {
      break;
    }

    for (let i = top; i < bottom + 1; ++i) {
      res[i][right] = val++;
    }
    if (--right < left) {
      break;
    }

    for (let i = right; i > left - 1; --i) {
      res[bottom][i] = val++;
    }
    if (--bottom < top) {
      break;
    }

    for (let i = bottom; i > top - 1; --i) {
      res[i][left] = val++;
    }
    if (++left > right) {
      break;
    }
  }
  return res;
};
// @lc code=end

File: 59.spiral-matrix-ii.py
#
# @lc app=leetcode id=59 lang=python3
#
# [59] Spiral Matrix II
#

# Time:  O(n^2)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        res, val = [[0] * n for _ in range(n)], 1
        left, right, top, bottom = 0, n - 1, 0, n - 1
        while left <= right and top <= bottom:
            for j in range(left, right + 1):
                res[top][j] = val
                val += 1
            top += 1
            if top > bottom:
                break

            for i in range(top, bottom + 1):
                res[i][right] = val
                val += 1
            right -= 1
            if right < left:
                break

            for j in range(right, left - 1, -1):
                res[bottom][j] = val
                val += 1
            bottom -= 1
            if bottom < top:
                break

            for i in range(bottom, top - 1, -1):
                res[i][left] = val
                val += 1
            left += 1
            if left > right:
                break

        return res


# @lc code=end
